Add two FX chain with Chorus, Delay, Reverb effects

soyersoyer opened the pull request:
new parameters:

FX[nFX]Send[nTG] (0..99) default: 50
FX[nFX]ChorusEnable1 default: 0
FX[nFX]ChorusEnable2 default: 0
FX[nFX]ChorusLFORate1 default: 50
FX[nFX]ChorusLFORate2 default: 83
FX[nFX]DelayTimeL: default: 36 (0.36s)
FX[nFX]DelayTimeR: default: 36 (0.36s)
FX[nFX]DelayFeedback: default: 60
FX[nFX]DelayTone: default: 80
FX[nFX]DelayPingPong: default: 0
FX[nFX]DelayMix: default: 0
FX[nFX]ReverbEnable: 0
FX[nFX]ReverbSize: default: 70
FX[nFX]ReverbHighDamp default:50
FX[nFX]ReverbLowDamp default: 50
FX[nFX]ReverbLowPass default: 30
FX[nFX]ReverbDiffusion: default: 65
FX[nFX]Level: default: 0

```
TG[1-8] -> EQ[1-8] -> Compressor[1-8] ---> TGMixer --------------------------------------------------+-> MasterEQ -> Limiter --> MasterVolume
                                      |     (Pan)     	                                             |
                                      |                                                              |
                                      |-> FXMixer[1-2] -> Chorus[1-2] -> Delay[1-2] -> Reverb[1-2] --|
                                          (FxSend+Pan)
```

soyersoyer commented:
That's better than I thought :)

One idea:
Change TGMixer to FXMixer to have effects and level control

Banana71 commented:
Wow, that's coming much faster than expected.

- The chorus is probably a Roland 80s chorus from the Juno, so you can't go wrong with that :-)

- I'm familiar with the delay from other Raspberry projects, and unfortunately, it has flaws in the echo and feedback. If feedback is set to 0, the first echo should still come back. I'd like to be able to set a delay like a stereo echo.

- The reverb sounds familiar. The default values still need to be adjusted.

But you managed to get it right the first time without any crackling or audio noise, respect. My ears thank you.

Are the effects from the Zynthian project?

Banana71 commented:
> That sounds better than I thought :)

It sounds great. :-)

soyersoyer commented:
> I'm familiar with the delay from other Raspberry projects, and unfortunately, it has flaws in the echo and feedback. If feedback is set to 0, the first echo should still come back. I'd like to be able to set a delay like a stereo echo.

I fixed the first echo problem. And I also added a FeedbackOnly parameter to keep the old behaviour.
And the time can also be set for both channels.

The Reverb paramaters have been adjusted.

> Are the effects from the Zynthian project?

No, they are are ported from jnonis/fx-full.

> I see you're taking the reverb parameters from the performances and assigning them to the FX1 reverb. The FX1 send parameter isn't being transferred from the old reverb send.

Nice catch! It was set by ReverbSend, but was later overridden by the default value of Fx1Send.

Banana71 commented:
- Reverb send is now correctly applied, and the performances sound exactly the same on the DreamDexed as they do on the MiniDexed.


- I'm not quite sure about the delay yet.
  - There are three adjustable times. Changes to the time also override Time Left and Time Right. Is this intended?
  - `Time=0` sets the delay time to the MAX value.
  - When `PingPong=On` and `Feedback=0`, an echo is only played on the right side. I would expect an echo to be heard first on one side and then on the other.
  - A "Dream" would be to calculate the delay times based on the tempo. That would save a lot of work. Or perhaps, if possible, the delay in FX1 works as before with absolute times (1 - 1000 ms), and the delay times in FX2 are calculated based on the tempo?

```
Tempo=120 bpm
Time Left=1/4 #(1/32, 1/16, 1/8, 1/4, 1/2, 1/1)
Time Right=1/8
```

_Please don't be surprised that I'm so fussy and question everything. In my experience, the questions and comments come later. It's a bit uncomfortable for me._

Banana71 commented:
The reverb sounds in the viedeos really good, perfect for ambient music and more. I'm excited to test this

Banana71 commented:
**CloudSeed:**
- Over 40 parameters for a synthesizer that already has far too many parameters is basically a no-go. Thanks to Presents, it would be doable again. Unfortunately, all CloudSeed effects crackle more or less on my PI 4.

**Mix Dry/Wet:**
- Functionally, it's so good and covers everything. However, the sum of the individual values for a ratio control should always add up to 100.
- Several examples for the display of dry to wet: 100:0 - 90:10 - 80:20 - 50:50 - 35:65 - 17:83 - 10:90 - 0:100
- All values between 50 and 100 are equally loud. Values below 50 reduce the volume.

**Effect Chains:**
- I think this is all way too much and too complicated.
- Every TG should really be able to send to up to four effect chains.
- I find 8 TGs with two effects chains more than sufficient and saves resources. The sound also suffers from so many effects, and it quickly becomes bass-heavy and muffled. Or is this a test (8 TGs with 4 effects chains) for the next stage of expansion?
- I can't say what the best solution is. Is a virtual mixer with limited effects sends and a logical setup perhaps the simplest and easiest to communicate setup?

_I'm currently recreating the piano sound from Keane's "Everybody's Changing" (001:010) and it's really fun. The EQs and compressors are working at their limits :-) A bit of chorus and a very short delay. CloudSeed also works better with only 5 TGs._

soyersoyer commented:
> Over 40 parameters for a synthesizer that already has far too many parameters is basically a no-go. Thanks to Presents, it would be doable again. 

https://www.youtube.com/watch?v=k0fVcXn0yBM

The contols look interesting :) This plugin is quite good, it might be worth adding the controls. It's a lot of work though.

> Unfortunately, all CloudSeed effects crackle more or less on my PI 4.

Try it now, it's faster a bit.

> Several examples for the display of dry to wet: 100:0 - 90:10 - 80:20 - 50:50 - 35:65 - 17:83 - 10:90 - 0:100

However, this hides the real values and one can only guess how it works. The other one is clear based on the display. (at least for me :)

This is more like two encoders mixed together, not a ratio.

> Or is this a test (8 TGs with 4 effects chains) for the next stage of expansion?

I thought of the 4FX chain more as an experience.
That way I can try out the dry/wet control, so that the TG doesn't sound fixed (only FX1 used).
Or also that the TGs are on FX1, on FX2 and FX3 the effects.

I saw in the MIDI CC list that there are 5 effects with controls
91 Effect 1 Depth 0-127 Usually controls reverb send amount
92 Effect 2 Depth 0-127 Usually controls tremolo amount
93 Effect 3 Depth 0-127 Usually controls chorus amount
94 Effect 4 Depth 0-127 Usually controls detune amount
95 Effect 5 Depth 0-127 Usually controls phaser amount

91 and 92 work in principle :)

I don't see the best solution either.
Expanding this to 4 parts is relatively easy.
But it might be more logical to have another layer.

> I'm currently recreating the piano sound from Keane's "Everybody's Changing"
I'm looking forward to it :)

Performances that are not compatible with MiniDexed should be moved to another branch or repo so that the Minidexed build continues to be good.

Banana71 commented:
Which version of CloudSeed is on the DreamDexed, version 1 or 2?

Cloud Seed is a powerful tool, but I believe it’s sufficient if a reverb appears only once in an effects chain.
Stacking multiple reverbs in the same chain is usually unnecessary and, in most cases, doesn’t add musical value.

Why?

Resource efficiency: Reverbs are some of the most CPU-intensive effects. Using more than one in a single chain quickly becomes wasteful.

Clarity: One well-tuned reverb is easier to balance than two competing tails.

Workflow: Too many overlapping reverbs create parameter overload and make the chain harder to control.

Studio practice: In most studios you wouldn’t stack multiple reverbs in one serial chain. Typically you choose the one that fits best, or use different reverbs on separate sends (e.g. a short room and a long hall).


```ruby
FX1 – Classic: [Chorus] → [Delay | Mix] → [Classic Reverb | Mix]

FX2 – Ambient: [Chorus] → [Delay | Mix] → [Cloud Seed (Long Reverb) | Mix]

FX3 – Motion: [Stereo Enhancer | Mix] → [Auto-Wah / Envelope Filter | Mix]
          → [Phaser / Resonant Filter | Mix] → [Short Reverb / Delay | Mix]

FX4 – Heavy: [Distortion | Mix] → [Chorus / Phaser / Flanger | Mix] 
          → [Reverb / Delay | Mix] → [Tremolo | Mix]
```

Please always question my ideas; I'm just throwing out whatever's on my mind.


I won't publish performances for the DreamDexed until Stage 2 is finished, and then in a separate folder or completely in another repo.

soyersoyer commented:
> Which version of CloudSeed is on the DreamDexed, version 1 or 2?

The 2.  The [core](https://github.com/GhostNoteAudio/CloudSeedCore) is open-source.

> Cloud Seed is a powerful tool, but I believe it’s sufficient if a reverb appears only once in an effects chain.
Stacking multiple reverbs in the same chain is usually unnecessary and, in most cases, doesn’t add musical value.

You don't have to enable the reverb before :)
The effects are in a fixed place in the menu because it seemed the easiest. But that will change soon.

>(If I remember correctly, the FX Full version of Jnonis included a very good reverb that produced very usable results (Ambient). I don't remember if it was the TAL Reverb or the Plate Reverb.)

The plate reverb is the default 'reverb' in MiniDexed, so it was probably the TAL reverb.
It only has 7 parameters and is probably not that CPU intensive. I'll try to integrate it.

Does the studio practice you wrote apply to one instrument or to more than one?

Banana71 commented:
> You don't have to enable the reverb before :)
The effects are in a fixed place in the menu because it seemed the easiest. But that will change soon.

> Does the studio practice you wrote apply to one instrument or to more than one?

I just wanted to say that multiple reverbs in one track are rather unusual. An instrument will usually only have one reverb effect. Of course, an entire song (project) can have several different reverbs.
But yes, of course, it's up to you whether you activate multiple reverbs. 
```swift
FX: [Chorus] → [Delay | Mix] → [Classic Reverb | Mix] → [Cloud Seed (Long Reverb) | Mix] → Master
```

Thanks to you, MiniDexed, DreamDexed, has never sounded as good as it does now.

soyersoyer commented:
This is only a rebase on the main (circle50, gcc14, security fixes, safety first:) )

> Thanks to you, MiniDexed, DreamDexed, has never sounded as good as it does now.

I had some WOW moments also, it was hard to stop playing :)

Banana71 commented:
The silent switching works as long as no delay is used. There's no delay reset.

soyersoyer commented:
Meanwhile I realized what's bothering me about the Master-FX1-FX2 setup. I can't get the FX-send higher than 100%. If the FX-send were 0-127 (0-> 0%; 64-> 100%; 127->200%), that would be fine for me. By the way, regular mixers work similarly. What do you think about this idea?

soyersoyer commented:
> To be honest, I don't understand what bothered you about it so far. Can you give me an example?

Sometimes I want the FX section (now FX2) to be louder than the master (now FX1). 
For example, the Rock Organ with the CloudSeed Ghostly preset sounds better in my opinion if the FX1 (master) level is 70 and the FX2 (CloudSeed) is 100.
This could not be achieved with the original Master-FX1-FX2 setup, which only allows the FX level to be set from 0-100%.
A similar thing could be achieved with CloudSeed's Dry/Early Wet/Late Wet controls (If they could be adjusted).

>  Doesn't an FX send of 200% pose new problems, like clipping?

Clipping is also possible now, because the reverb is added to 100% of the master.

> For me, there's no such thing as an FX send in the traditional sense; it's a volume control that adjusts the signal from the TG (synth) to the effects processor.

Yes, but the volume can be increased and decreased from the base level, not just decreased.

> Now that you've made me rethink the concept, I've realized that our menu order has little to do with the signal path.

Are you thinking that the order is not entirely logical?

> Could this be related to your problem?

No, but it's okay to reconsider :)

> Audio Signal Path:

Synth_Dexed Pluginfx (DC filter - Volume - Cutoff - Resonance) -EQ - Compressor - Pan+FX-Send

soyersoyer commented:
Limiter -> Master Compressor

Banana71 commented:
> Sometimes I want the FX section (now FX2) to be louder than the master (now FX1).
> For example, the Rock Organ with the CloudSeed Ghostly preset sounds better in my opinion if the FX1 (master) level is 70 and the FX2 (CloudSeed) is 100.
> This could not be achieved with the original Master-FX1-FX2 setup, which only allows the FX level to be set from 0-100%.
> A similar thing could be achieved with CloudSeed's Dry/Early Wet/Late Wet controls (If they could be adjusted).

I still don't fully understand the problem. You split the effects send into two paths (FX1 + FX2) and want to achieve a higher proportion of the effects path with CloudSeed (FX2)?
I believe that a sound like Rock Organ only needs one effects send, and each individual effect in the chain (chorus, delay, reverb (Classic or CloudSeed)) can adjust the effect level with a wet/dry control. 
Wouldn't it be enough to simply add a wet/dry control to CloudSeed, just like you did with the chorus?

---

> Are you thinking that the order is not entirely logical?

Yes, exactly. I think it would be easier for many users if the menu structure followed the logical progression of the sound engine. But I would lower the priority for now.

```
TG Synth Part:
Voice - Bank - Detune - Pitch-Bend - Portamento - Note-Limit - Poly - Modulation - Channel -  Edit Voice - TG-Link

TG Audio Signal Path:
DC Filter - Volume - Cutoff + Resonance - EQ - Compressor - Pan - FX-Send (Volume control for the virtual effects device)
```


---
I've stored some performances in Bank 5 to test the DreamDexed; you might be interested in them, too.
[005_DreamDexed.zip](https://github.com/user-attachments/files/22439368/005_DreamDexed.zip)

soyersoyer commented:
> You split the effects send into two paths (FX1 + FX2) and want to achieve a higher proportion of the effects path with CloudSeed (FX2)?

Yes.

> Wouldn't it be enough to simply add a wet/dry control to CloudSeed, just like you did with the chorus?

Yes, that can work too. I will do it soon, and setting the other parameters wouldn't be bad either.

Actually, what bothers me is that on a classic mixer the level can be set between -∞, -15 dB .. 0 .. +15 dB, while here it is between 0-100%.
If I want a little more from one FX chain and it's already at 100%, for example, I can only do that by turning down another one.
So, since there are multiple chains, it's not a problem.
However, if the volume of the FX chain could be adjusted similarly to a classic mixer, then one less FX chain would be enough for the same thing, and the TGs could be mixed into the master without an FX chain.
Maybe because of the similarity to classic mixers people would understand it easier.
However, the current solution, where everything goes through an FX chain, is more flexible.
But I'm not sure if this flexibility is really needed.
What do you think?

> I've stored some performances in Bank 5 to test the DreamDexed; you might be interested in them, too.

Wow, these are very cool! Thanks!

soyersoyer commented:
I limited the length of the lines when writing to the LCD, so it will truncate them if they are longer.
The Compressor and TG Mixer don't fit. I don't know if it's worth shortening them.

FXn-Sends have been moved to the TG menu. And they are working fine now.

soyersoyer commented:
Forgot to rename FXOutputLevel...

Banana71 commented:
My RPi Zero 2W now has a heatsink, and `fast=true` is entered in the cmdline.txt file.
```
//usbspeed=full
fast=true
```
Unfortunately, there are still crackling noises when a compressor is switched on. But with a little care, very good sounds are possible with 8TGs and CloudSeed2 on the RPi Zero 2W. Certainly, the priority should be to move the project forward and then later see what can be done to get it running on the weaker RPi models as well.

The latest version looks and sounds really good. 👍

Banana71 commented:
I have been intensively comparing the MiniDexed compressor (limiter) with the DreamDexed master compressor. When the climiter in MiniDexed engages, it does indeed change the sound, as you suspected. With the 001:030_Petersdom performance, you can hear very clearly that the compressor primarily reduces the bass frequencies.

In DreamDexed, the master compressor, with the selected default parameters, is too aggressive.
You have different default parameter sets for the master compressor in DreamDexed, depending on whether a MiniDexed performance has Compressor `Enable=0` or `Enable=1`. Could you please use the parameter sets from `Enable=0` in both cases:
```
Pre Gain=0
Threshold=-3
Ratio=20:1
Attack=5ms
Release=5ms
HPfilter=On
```
This way, the difference won't be so noticeable.

Is this still the current code for the compressor in MiniDexed (Synth_Dexed)? [MiniDexed Issues #7](https://github.com/probonopd/MiniDexed/issues/7)
If so, we could consider adding this limiter to DreamDexed as a "Vintage Limiter," placing it after the master compressor, and disabling the master compressor by default for MiniDexed performances. The code was created because the standard compressor was putting too much strain on the CPU.!?
